Add git-backed server mode for editing

jamesgorman2 opened the pull request:
Based on TiddlyWiki server using `simple-git`.

serra commented:
@jamesgorman2 I have tested this setup and it works on my machine as well.

One concern: I started out on a local branch without a remote tracking branch set up. I then made some edits to an existing tiddler and hit save. This resulted in an obscure error message (sorry forgot to copy) in the browser and a loss of work - I could not retrieve the changes I made.

Can we think of more error scenarios like this? I think we should really make this safe for contributions. And loosing work feels very unsafe.

I also found that this setup commits and pushes files outside the tiddlers directory (e.g. see b1778ed94472140cc3d7f5a5f1d96b27bf544555), which I find strange.

jamesgorman2 commented:
- [ ] only commit/push `./tiddlers`
- [ ] update readme to match how to tiddler
- [ ] update recommended exec to emit git debug messages

> @jamesgorman2 I have tested this setup and it works on my machine as well.
> 
> One concern: I started out on a local branch without a remote tracking branch set up. I then made some edits to an existing tiddler and hit save. This resulted in an obscure error message (sorry forgot to copy) in the browser and a loss of work - I could not retrieve the changes I made.
> Can we think of more error scenarios like this? I think we should really make this safe for contributions. And loosing work feels very unsafe.
> 

Git is a dreadful programming language, so there's not a lot I can do about this stuff without sinking more time into it than it's worth. Most users will be coming in from the how to, so I'd rather fix it in documentation than code.

Also, is the missing work in a stash? It should be very hard to lose things forever. If you run debug you can check what's happening in git

> I also found that this setup commits and pushes files outside the tiddlers directory (e.g. see [b1778ed](https://github.com/zefram-cochrane/Zefram-Cochrane.github.io/commit/b1778ed94472140cc3d7f5a5f1d96b27bf544555)), which I find strange.

This was me not sure of what gets updated by TW, should be an easy change.

serra commented:
Is our goal to make it more easy for non-git savvy people to contribute? If so then I think emitting git debug messages is not really helpful.

I'll check again to see if I can find a simple way to lose changes and check back here.

serra commented:
To reproduce a lost data scenario:

```
[fail] Scenario: no remote tracking branch
v Given I have created a local branch
v   and I have not created a remote tracking branch
v   and I have made some changes
v  when I click save
x  then I get a friendly error message  
x   and my changes are not lost
x   and my changes are visible to me
```

See details below, including `git status` and `git stash show`. To me, my edits are lost.

<details>
<pre>
~$ cd /Users/marijn/dev/Zefram\-Cochrane
Zefram-Cochrane$ DEBUG=* npm run tiddlywiki -- . --gitserver

> @ tiddlywiki /Users/marijn/dev/Zefram-Cochrane
> tiddlywiki "." "--gitserver"

GitServer!
Serving on 127.0.0.1:8080
(press ctrl-C to exit)
 syncer-server-filesystem: Dispatching 'save' task: $:/StoryList 
 filesystem: Saved file /Users/marijn/dev/Zefram-Cochrane/tiddlers/$__StoryList.tid 
 syncer-server-filesystem: Dispatching 'save' task: /Users/marijn/dev/Zefram-Cochrane/tiddlers/favicon.ico 
 filesystem: Saved file /Users/marijn/dev/Zefram-Cochrane/tiddlers/favicon.ico 
 syncer-server-filesystem: Dispatching 'save' task: $:/StoryList 
 filesystem: Saved file /Users/marijn/dev/Zefram-Cochrane/tiddlers/$__StoryList.tid 
 syncer-server-filesystem: Dispatching 'save' task: $:/plugins/felixhayashi/tiddlymap/misc/defaultViewHolder 
 filesystem: Saved file /Users/marijn/dev/Zefram-Cochrane/tiddlers/$__plugins_felixhayashi_tiddlymap_misc_defaultViewHolder.tid 
 syncer-server-filesystem: Dispatching 'save' task: $:/StoryList 
 filesystem: Saved file /Users/marijn/dev/Zefram-Cochrane/tiddlers/$__StoryList.tid 
 syncer-server-filesystem: Dispatching 'save' task: $:/StoryList 
 filesystem: Saved file /Users/marijn/dev/Zefram-Cochrane/tiddlers/$__StoryList.tid 
 syncer-server-filesystem: Dispatching 'save' task: Draft of 'Contributors' 
For Draft of 'Contributors', type is application/x-tiddler hasMetaFile is false filepath is /Users/marijn/dev/Zefram-Cochrane/tiddlers/Draft of 'Contributors'.tid
 filesystem: Saved file /Users/marijn/dev/Zefram-Cochrane/tiddlers/Draft of 'Contributors'.tid 
 syncer-server-filesystem: Dispatching 'save' task: Draft of 'Contributors' 
 filesystem: Saved file /Users/marijn/dev/Zefram-Cochrane/tiddlers/Draft of 'Contributors'.tid 
 syncer-server-filesystem: Dispatching 'delete' task: Draft of 'Contributors' 
 filesystem: Deleted file /Users/marijn/dev/Zefram-Cochrane/tiddlers/Draft of 'Contributors'.tid 
 syncer-server-filesystem: Dispatching 'save' task: $:/StoryList 
 filesystem: Saved file /Users/marijn/dev/Zefram-Cochrane/tiddlers/$__StoryList.tid 
 syncer-server-filesystem: Dispatching 'save' task: Contributors 
 filesystem: Saved file /Users/marijn/dev/Zefram-Cochrane/tiddlers/Contributors.tid 
Assert is repo...
  simple-git [ 'rev-parse', '--is-inside-work-tree' ] +0ms
Assert named editor...
  simple-git [ 'config', '--global', 'user.name' ] +30ms
Revert unwanted changes...
  simple-git [ 'status', '--porcelain', '-b', '-u' ] +26ms
  simple-git [ 'diff', '--', 'tiddlers/Contributors.tid' ] +34ms
  simple-git [ 'status', '--porcelain', '-b', '-u' ] +26ms
Merge upstream...
  simple-git [ 'status', '--porcelain', '-b', '-u' ] +24ms
  simple-git [ 'stash', '--include-untracked' ] +22ms
  simple-git [ 'pull' ] +358ms
There is no tracking information for the current branch.
Please specify which branch you want to merge with.
See git-pull(1) for details.

    git pull <remote> <branch>

If you wish to set tracking information for this branch you can do so with:

    git branch --set-upstream-to=<remote>/<branch> gitserver-branch-wo-tracking-branch


Resolving conflicts...
  simple-git [ 'status', '--porcelain', '-b', '-u' ] +608ms
  simple-git [ 'reset', '--soft' ] +20ms
  simple-git [ 'stash', 'drop' ] +19ms
Error saving: [object Object]
^C
Zefram-Cochrane$ git status
On branch gitserver-branch-wo-tracking-branch
nothing to commit, working tree clean
Zefram-Cochrane$ git stash show
No stash entries found.
</pre>

</details>

serra committed:
Add a contribution banner

Todo: add some styling to make it stand out more.

serra committed:
Add stylesheet for contribution banner

serra committed:
Change text of contribution banner

serra commented:
Hi James, I tested the above scenario with your changes from 8538995. On my machine it fails in exactly the same way as before.

serra commented:
I did another test. This one is more easy to reproduce, it will occur more often and I suspect it touches the same issue:

```
[fail] Scenario: no connection to remote tracking branch
v Given I have created a local branch
v   and I have created a remote tracking branch
v   and I have made some changes
v   and I disconnect from the internet so I'm no longer connected to my remote tracking branch
v  when I click save
x  then I get a friendly error message  
x   and my changes are not lost
x   and my changes are visible to me
```

serra commented:
I'm not aware of the internals, but would it be an option to commit the changes locally before doing anything with our remote branches? That way our changes would always be safe. Or am I being naive here?

serra commented:
Hi @jamesgorman2, on [mergeUpstream.js#L107](https://github.com/zefram-cochrane/Zefram-Cochrane.github.io/blob/85389952aafb716cbd3d2842b3e43d1ef296f1c9/plugins/gitserver/git/mergeUpstream.js#L107), you do:

```
    .then(() => git.stash(['drop']))
```

This basically sends all our changes to oblivion.

Wouldn't simply changing this to:
```
    .then(() => git.stash(['pop']))
```

resolve our problems? then it would simply reapply the stash to our current working directory and we'd at least still have our changes.

serra commented:
Hm, while the change I suggested does fix the above scenarios, there is probably some other scenarios it breaks. Do we have some tests for those @jamesgorman2; e.g. for merge conflicts with our tracking branch.

jamesgorman2 commented:
I hadn't looked at the remote tracking being missing as to make the workflow work, it requires you to be pushing to master (so you can get a meaningful merge). That said, it shouldn't break everything if the remote is unavailable. I'll have a look at that.

It is possible to get rid of the 'drop', though if the code reaches here a 'pop' has already been attempted. If this continues to be problematic then the only real option is to remove the 'drop' and fill up the user's stash stack.

The flow is:
* assertions
* stash --include-untracked
* pull
* stash pop
* if conflicts: (this path is currently taken for _all_ exception after the stash command)
    * copy conflict files to draft files, reset to HEAD
    * reset --soft (clean up the merge state)
    * stash drop
* else:
    * update editors
    * generate index.html
    * commit
    * push

serra commented:
Great @jamesgorman2 . I think graceful handling of the scenario of an unavailable remote is sufficient at this stage.

jamesgorman2 commented:
These latest changes should
* add editors
* abort and restore changes for any failure that is not conflicts
* report proper messages

serra commented:
Will review this later today @jamesgorman2 !

serra commented:
@jamesgorman2 

I can confirm that this handles errors gracefully - I understand the error message and no work is lost 👍 

However, the following happy flow no longer works:

```
[fail] Scenario: happy flow
v Given I have created a local "master" branch
v   and I have created a remote tracking branch "master" on remote "origin"
v   and I have made some changes
v  when I click save
x  then my changes get committed into my local git repository 
x   and my changes are pushed to origin
```

jamesgorman2 commented:
Thanks Marijn. It helps if I remember to uncomment the commit and push code.

serra commented:
```
[success] Scenario: happy flow
v Given I have created a local "master" branch
v   and I have created a remote tracking branch "master" on remote "origin"
v   and I have made some changes
v  when I click save
v  then my changes get committed into my local git repository 
v   and my changes are pushed to origin

[success] Scenario: no connection to remote tracking branch
v Given I have created a local branch
v   and I have created a remote tracking branch
v   and I have made some changes
v   and I disconnect from the internet so I'm no longer connected to my remote tracking branch
v  when I click save
v  then I get a friendly error message  
v   and my changes are not lost
v   and my changes are visible to me
v   and I can save my changes after reconnecting
```

serra commented:
@jamesgorman2 :tada: let's ship this!

zefram-cochrane commented:
Thanks for the lovely work fellows. I've got some git-ignoramus questions:

1) Simple sequence of commands so I can run this myself and see it working?
2) When I merge the pull request should I squash and merge or rebase and merge?

jamesgorman2 commented:
To get things running on a test branch you need to:
* fetch from my fork
* create a branch in your repo to run against (so that changes can be saved)
* merge my changes into that branch
* build and run the app 

I think this will work, ping me if any of the steps fail.

```
# if you don't have my fork as a remote
git remote add jamesgorman2 git@github.com:jamesgorman2/Zefram-Cochrane.github.io.git
# fetch remotes
git fetch --all
# get your own branch so that it will push changes
git checkout master
git checkout -b pull6
# merge in my changes
git merge jamesgorman2/master
# push remote so it can save changes
git push --set-upstream origin pull6

# build app
npm install
# run app with git debug logging
DEBUG=* npm run tiddlywiki -- . --gitserver
```

zefram-cochrane commented:
[Passepartout-1073 pete ~/src/Zefram-Cochrane.github.io  (jamesgorman2-master)] git remote add jamesgorman2 git@github.com:jamesgorman2/Zefram-Cochrane.github.io.git
[Passepartout-1073 pete ~/src/Zefram-Cochrane.github.io  (jamesgorman2-master)] git fetch --all
Fetching origin
Fetching jamesgorman2
Warning: Permanently added the RSA host key for IP address '192.30.255.113' to the list of known hosts.
Permission denied (publickey).
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
error: Could not fetch jamesgorman2
[Passepartout-1073 pete ~/src/Zefram-Cochrane.github.io  (jamesgorman2-master)]

serra commented:
@zefram-cochrane have you tried `git remote add james_over_https https://github.com/jamesgorman2/Zefram-Cochrane.github.io.git`?

jamesgorman2 commented:
Thanks @serra, I ended up chatting to Pete via Slack.

New version up that correctly deals with conflicts. What we're still missing:
* test if opening up the non-draft copy clobbers the draft. I thought I saw this happening, but want to check again after cleaning things out
* a way to update the wiki after the save - currently this requires at least refresh, maybe restarting the application
* a button to just do an update from the remote - there is a stub but I haven't wired it up in the UI (since I'm not sure how).

serra commented:
Those are good points @jamesgorman2 . Do we _really_ need them before sharing the work so far?

At this stage, contributor's edits are safe; at least so it seems so to me. If they run into problems, we can help them out with; reverting to commandline git if needed.

jamesgorman2 commented:
My update to this is that it works _so long as there are no conflicts_. If there are conflicts things seem to be getting weird with files that should have been reset appearing in git as conflicted. I don't know why this is happening and will have to spend a while investigating. Beyond that, it is ready to use (at least on mac)

jamesgorman2 commented:
I also have to work out what to do when headers conflict. I wasn't testing for this. eg

```
-created: 20181111055700759
-list: [[XSCALE Practices]] [[XSCALE Structures]] [[XSCALE Metrics]] [[XSCALE Principles]] [[Welcome Visitors]]
-modified: 20181111055700759
+editors: [{"name":"James Gorman","timestamp":1553074528681}]
+created: 20190320093517379
+list: [[New Tiddler]] [[XSCALE Practices]] [[XSCALE Structures]] [[XSCALE Metrics]] [[XSCALE Principles]] [[Welcome Visitors]]
+modified: 20190320093517379
```

Not actually sure how to deal with this since it is possible the conflict markers could go over the header break.

serra commented:
@jamesgorman2 maybe we could open it up to a select number of coaches.
I was thinking to announce on Slack that we will open the wiki for a couple of coaches who want to collaborate using a rough-around-the-edges version. Coaches can opt-in participating by adding their name to the contributors page through a pull request; they can do this [through the Github UI](https://github.com/zefram-cochrane/Zefram-Cochrane.github.io/blob/master/tiddlers/Contributors.tid). To do this they need a github account. Once we receive such a PR, we merge and give this user push access to the repo.

serra commented:
E.g. invite them to participate on content on the guide.

serra commented:
@jamesgorman2 IIRC there is an option to disable tracking information in the headers. E.g. you can turn of tracking of created/modified timestamps. Would that reduce the chance of conflicts?

jamesgorman2 commented:
My main concern is that it risks being excessively arcane for the majority of users (whether the direct edit mode currently under test, or your PR based proposal). The behaviour is sufficiently complex _and_ complicated that I'm having difficulty thinking about.

More directly disabling tracking will reduce the number of header conflicts (it already ignores timestamp only changes), but well lose the recent changes list.

I've asked Pete to jot down what his top line requirements are and want to put together a quick decision matrix to see if this is the right path or we should look at something different.

serra committed:
Add simple build script

serra committed:
Simplify build script

I wonder if this uses the correct node version

serra committed:
Publish `./output/index.html` to gh_pages branch

serra committed:
Use lts node version

jamesgorman2 committed:
Hide save and delete buttons

jamesgorman2 committed:
Update Contribution Banner

serra committed:
Merge pull request #2 from serra/build-script

Add simple build script

serra committed:
Add build badge and remove zefram from title

jamesgorman2 committed:
Merge remote-tracking branch 'xscale/master' into contribution-banner

jamesgorman2 committed:
Update git path

jamesgorman2 committed:
Update contribution message

serra committed:
Save an offline version in the build

This, way the server side components do not get included.

serra committed:
Merge pull request #4 from jamesgorman2/contribution-banner

Use github contributions

zefram-cochrane committed:
Haha, I have scragged the metadata!

Hoping the autobuild stops me from being able to check this in ...

zefram-cochrane committed:
Unscragging ...

jamesgorman2 committed:
Basic test for correct metadata

jamesgorman2 committed:
Update build and README

jamesgorman2 committed:
Remove drafts

jamesgorman2 committed:
Update editing the wiki

jamesgorman2 committed:
Typo

jamesgorman2 committed:
Update with adding new tiddlers
